### PR TITLE
[WFCORE-3596] Introduce named permission sets in the Elytron subsystem

### PR DESCRIPTION
**Pull request by fjuma**
https://issues.jboss.org/browse/WFCORE-3596
https://issues.jboss.org/browse/WFWIP-9
https://issues.jboss.org/browse/EAP7-978

Analysis document: https://github.com/wildfly/wildfly-proposals/pull/43

Capabilities PR: https://github.com/wildfly/wildfly-capabilities/pull/23

Documentation and test fix PR: https://github.com/wildfly/wildfly/pull/11194

WARNING: After this PR is merged, the above WildFly PR will need to be merged to prevent a test case that is no longer valid from failing.

The following experimental job shows the WildFly testsuite passing with the above WildFly changes:

 https://ci.wildfly.org/viewLog.html?buildId=102976&tab=buildResultsDiv&buildTypeId=WF_WildFlyCoreIntegrationExperiments

### CODE REVIEW

**Comment by kabir** [comment]
@bstansberry @jmesnil All the premerge requirements are done.

@fjuma I have merged the upstream PR (although it includes docs for something not merged in full yet :) )

**Comment by bstansberry** [comment]
@aloubyansky can you fpaste or something an example of what a feature-group-spec (e.g. ejb3.xml) that needs to insert into the permission set would look like?

I also think there will be some gymnastics needs getting this merged in coordination with the galleon PR in full.  Probably best is this goes first and the galleon PR then adapts to the new elytron configs.

File: elytron/src/main/resources/subsystem-templates/elytron.xml [hunk 1]
```diff
@@ -60,6 +60,15 @@
...
         </mappers>
+        
+        <permission-sets>
+            <permission-set name="login-permission">
+                <permission class-name="org.wildfly.security.auth.permission.LoginPermission" />
+            </permission-set>
+            <permission-set name="default-permissions">
+                <?ADDITIONAL_PERMISSIONS?>
+            </permission-set>
+        </permission-sets>
```
[the remark this answers is not in the document]
Comment (fjuma) [reply elytron/src/main/resources/subsystem-templates/elytron.xml]: I've now updated elytron-common.xml in the core-galleon-pack to add in the login-permission and default-permissions permission sets. I've also updated the spec for the default-permission-mapper to reference these permission sets. @aloubyansky please review.

**Comment by darranl** [comment]
I would recommend we get these changes in first as well, the primary motivation for these changes was to provide named resources for Gelleon to be able to operate on so if we get this in first if there are any workarounds for the prior default config those can be removed.

**Comment by aloubyansky** [comment]
@fjuma you got it right, thanks! If that's ok I'd like to include a couple of additional changes here.
There is a feature group host-elytron-common.xml which is referenced from the host-profile.xml. I'm not sure why we ended up with host-elytron-common.xml, it appears to be a copy of elytron-common.xml. Could you please remove host-elytron-common.xml and adjust host-profile to reference elytron-common? Like this

> diff --git a/core-galleon-pack/src/main/resources/feature_groups/host-profile.xml b/core-galleon-pack/src/main/resources/feature_groups/host-profile.xml
> index a47333b033..53ad292d26 100644
> --- a/core-galleon-pack/src/main/resources/feature_groups/host-profile.xml
> +++ b/core-galleon-pack/src/main/resources/feature_groups/host-profile.xml
> @@ -4,7 +4,7 @@
>      <feature spec="subsystem.elytron">
>          <param name="final-providers" value="combined-providers"/>
>          <param name="disallowed-providers" value="[OracleUcrypto]"/>
> -        <feature-group name="host-elytron-common">
> +        <feature-group name="elytron-common">
>              <include feature-id="subsystem.elytron.file-audit-log:file-audit-log=local-audit">

Otherwise, we need to apply the same changes to host-elytron-common to make the host configs match their corresponding legacy versions. Thanks!

**Comment by fjuma** [comment]
@aloubyansky I've removed host-elytron-common.xml and adjusted host-profile.xml accordingly.

**Comment by aloubyansky** [comment]
@fjuma thanks!

**Comment by aloubyansky** [comment]
An example of an EJB permission added to the default-permissions set. In the ejb group the following feature will need to be added
```
    <feature spec="subsystem.elytron.permission-set">
        <param name="permission-set" value="default-permissions"/>
        <feature spec="subsystem.elytron.permission-set.permissions">
            <param name="module" value="org.jboss.ejb-client"/>
            <param name="class-name" value="org.jboss.ejb.client.RemoteEJBPermission"/>
        </feature>
    </feature>

```
Note, it could also be configured using the complex attribute on the permission-set feature by specifying the value as the complete list.

This translates into the following operations
```
/subsystem=elytron/permission-set=default-permissions:add(permissions=[])
/subsystem=elytron/permission-set=default-permissions:list-add(name=permissions,value={"class-name" => "org.jboss.ejb.client.RemoteEJBPermis
sion","module" => "org.jboss.ejb-client"})

```

The result in the elytron subsystem config then looks like this
```
                <permission-set name="default-permissions">
                    <permission class-name="org.jboss.ejb.client.RemoteEJBPermission" module="org.jboss.ejb-client"/>
                </permission-set>

```

**Comment by bstansberry** [comment]
@aloubyansky Thanks.  So, here's the key magic, in case others happen to read this and want to understand.

The use of 'list-add' is due to the feature spec for subsystem.elytron.permission-set.permissions:

```
<feature-spec xmlns="urn:jboss:galleon:feature-spec:1.0" name="subsystem.elytron.permission-set.permissions">
    <annotation name="jboss-op">
        <elem name="name" value="list-add"/>
        <elem name="op-params" value="class-name,module,target-name,action"/>
        <elem name="complex-attribute" value="permissions"/>
        <elem name="addr-params" value="profile,host,subsystem,permission-set"/>
    </annotation>.
     . . 
</feature-spec>
```

That spec is, like all of the resource/attribute specs, generated from output from ReadFeatureDescriptionHandler.  The key bit there is currently like this:

```
   private void processListAttribute(final ModelNode parentFeature, final ImmutableManagementResourceRegistration registration, ObjectListAttributeDefinition objAttDef) {
        //System.out.println("List attr " + registration.getPathAddress().toCLIStyleString() + " " + objAttDef.getName());
        final ModelNode parentSpecName = parentFeature.require(NAME);
        ModelNode attFeature = new ModelNode();
        final String specName = parentSpecName.asString() + "." + objAttDef.getName();
        attFeature.get(NAME).set(specName);
        ModelNode annotation = attFeature.get(ANNOTATION);
        annotation.get(ModelDescriptionConstants.NAME).set("list-add");
        annotation.get(COMPLEX_ATTRIBUTE).set(objAttDef.getName());
        if (parentFeature.hasDefined(ANNOTATION)) {
            annotation.get(ADDR_PARAMS)
                    .set(parentFeature.require(ANNOTATION).require(ADDR_PARAMS));
            if (parentFeature.require(ANNOTATION).hasDefined(ADDR_PARAMS_MAPPING)) {
                annotation.get(ADDR_PARAMS)
                        .set(parentFeature.require(ANNOTATION).require(ADDR_PARAMS_MAPPING));
            }
        } else {
            addParams(attFeature, registration.getPathAddress(), new ModelNode().setEmptyList());
        }
```

**Comment by darranl** [comment]
@aloubyansky / @bstansberry One objective we wanted to achieve with the Elytron subsystem is that all resources can be defined as a single shot, with no need for batches / composite operations.

There is one exception to this where circular dependencies can be defined within the subsystem but other than that is should never be required that defining resource in the Elytron subsystem requires additional coordination.

IMO thinks like list-add should be a convenience operation for an administrator wishing to modify an existing resource but we shouldn't be using it for routine definitions.

It doesn't happen in this case but delaying adding permissions to a list add could actually lead to an identity being granted permissions they should not be granted for a short while until the next reload occurs.

**Comment by aloubyansky** [comment]
@darranl I am not sure what you mean by a single shot, if it's not a composite operation.

I don't think you should worry about list-add. Since this is simply a config generation, which is not happening at runtime and is hidden from the end user.

**Comment by darranl** [comment]
Yeah I am not worried that this would create the permission problem as it will never happen on a server that is up.

But it did make me think after Thursdays discussion where problems were encountered requiring multiple batches that maybe we should be minimising the operations actually required.

**Pull request merged**